pages/workoutlist.js

```diff
@@ -0,0 +1,176 @@
+import { useState } from "react";
+import styled from "styled-components";
+import { exercises } from "@/lib/exercises";
+import { workouts } from "@/lib/workouts";
+
+function findExerciseById(exerciseId) {
+  return exercises.find((exercise) => exercise.id === exerciseId);
+}
+
+export default function WorkoutsList() {
+  const [showDetails, setShowDetails] = useState({});
+
+  const toggleDetails = (workoutId) => {
+    setShowDetails((prev) => ({
+      ...prev,
+      [workoutId]: !prev[workoutId],
+    }));
+  };
+
+  return (
+    <>
+      <header>
+        <StickyHeader>Choose Your Workout</StickyHeader>
+      </header>
+
+      <Main>
+        <WorkoutList>
+          {workouts.map((workout) => {
+            const muscleGroupsSet = new Set();
+
+            workout.exercises.forEach((exercise) => {
+              const foundExercise = findExerciseById(exercise.exerciseId);
+
+              if (foundExercise) {
+                foundExercise.muscleGroups.forEach((muscleGroup) => {
+                  muscleGroupsSet.add(muscleGroup);
+                });
+              }
+            });
+
+            const reducedMuscleGroups = Array.from(muscleGroupsSet);
+            const isDetailsVisible = showDetails[workout.id];
+            return (
+              <WorkoutCard key={workout.id}>
+                <h2>{workout.name}</h2>
+                <SpotlightHeading>Muscles In The Spotlight:</SpotlightHeading>
+                <MuscleGroupList>
+                  {reducedMuscleGroups.map((muscle) => (
+                    <MuscleBadge key={muscle}>{muscle}</MuscleBadge>
+                  ))}
+                </MuscleGroupList>
+                <ToggleButton onClick={() => toggleDetails(workout.id)}>
+                  {isDetailsVisible ? "Show Less" : "Show More"}
+                </ToggleButton>
+                {isDetailsVisible && (
+                  <ExerciseList>
+                    {workout.exercises.map((exercise) => {
+                      const foundExercise = findExerciseById(
+                        exercise.exerciseId
+                      );
+
+                      return (
+                        <ExerciseItem key={exercise.exerciseId}>
+                          <ExerciseName>{foundExercise.name}</ExerciseName>
+                          <ExerciseDetails>
+                            Sets: {exercise.sets}
+                          </ExerciseDetails>
+                          <ExerciseDetails>
+                            Reps: {exercise.reps}
+                          </ExerciseDetails>
+                        </ExerciseItem>
+                      );
+                    })}
```

Comment:
To avoid extra work while rendering, create a precomputed structure before the return statement. For example, you can calculate the set of muscle groups outside of the map callback. This helps improve performance and makes the code easier to read by reducing the amount of work done during rendering.

Additionally, I recommend first storing all muscle groups in an array **and then** converting that array into a Set to ensure uniqueness. The new approach is more straightforward because it first collects all muscle groups into an array (`muscleGroups`). This step is easier to understand and manage since all operations are performed in a linear manner. After collecting all muscle groups, the array is converted into a Set to ensure uniqueness (`uniqueMuscleGroups`). This two-step process is more intuitive and easier to follow.
While both approaches are efficient, the suggested approach can be slightly more performant in certain scenarios, especially if the array of muscle groups is large and there are only a few unique values. By using a Set only once at the end, you avoid the repeated overhead of inserting elements into a Set multiple times.


```suggestion
const preparedWorkouts = workouts.map((workout) => {
    const muscleGroups = [];
    const enrichedExercises = workout.exercises.map((exercise) => {
      const foundExercise = findExerciseById(exercise.exerciseId);
      if (foundExercise) {
        muscleGroups.push(...foundExercise.muscleGroups);
        return { ...exercise, name: foundExercise.name };
      }
      return exercise;
    });

    const uniqueMuscleGroups = [...new Set(muscleGroups)];
    return {
      ...workout,
      exercises: enrichedExercises,
      muscleGroups: uniqueMuscleGroups,
    };
  });

  return (
    <WorkoutList>
      {preparedWorkouts.map((workout) => {
        const isDetailsVisible = visibleWorkouts.includes(workout.id);
        return (
          <WorkoutCard key={workout.id}>
            <h2>{workout.name}</h2>
            <SpotlightHeading>Muscles In The Spotlight:</SpotlightHeading>
            <MuscleGroupList>
              {workout.muscleGroups.map((muscle) => (
                <MuscleBadge key={muscle}>{muscle}</MuscleBadge>
              ))}
            </MuscleGroupList>
            <ToggleButton onClick={() => toggleDetails(workout.id)}>
              {isDetailsVisible ? "Show Less" : "Show More"}
            </ToggleButton>
            {isDetailsVisible && (
              <ExerciseList>
                {workout.exercises.map((exercise) => {
                  return (
                    <ExerciseItem key={exercise.exerciseId}>
                      <ExerciseName>{exercise.name}</ExerciseName>
                      <ExerciseDetails>Sets: {exercise.sets}</ExerciseDetails>
                      <ExerciseDetails>Reps: {exercise.reps}</ExerciseDetails>
                    </ExerciseItem>
                  );
                })}
              </ExerciseList>
            )}
          </WorkoutCard>
        );
      })}
    </WorkoutList>
  );
}
```